src/pu-archive.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ * Copyright (c) 2023 PHYTEC Messtechnik GmbH
+ */
+
+#include <archive.h>
+#include <archive_entry.h>
+#include <fcntl.h>
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include "pu-archive.h"
+#include "pu-error.h"
+#include "pu-utils.h"
+
+static gint
+pu_archive_copy_data(struct archive *read,
+                     struct archive *write)
+{
+    gint res;
+    gconstpointer buf;
+    gsize size;
+    goffset offset;
+
+    while (TRUE) {
+        res = archive_read_data_block(read, &buf, &size, &offset);
+        if (res == ARCHIVE_EOF)
+            return ARCHIVE_OK;
+        if (res < ARCHIVE_OK)
+            return res;
+
+        res = archive_write_data_block(write, buf, size, offset);
+        if (res < ARCHIVE_OK)
+            return res;
+    }
+}
+
+gboolean
+pu_archive_extract(const gchar *filename,
+                   const gchar *dest,
+                   GError **error)
+{
+
+    struct archive *ar;
+    struct archive *ext;
+    struct archive_entry *entry;
+    g_autofree gchar *old_dest = NULL;
+    gint flags;
+    gint read;
+    gint write;
+
+    flags = ARCHIVE_EXTRACT_TIME |
+            ARCHIVE_EXTRACT_PERM |
+            ARCHIVE_EXTRACT_OWNER |
+            ARCHIVE_EXTRACT_ACL |
+            ARCHIVE_EXTRACT_FFLAGS;
+
+    ar = archive_read_new();
+    archive_read_support_filter_all(ar);
+    archive_read_support_format_all(ar);
+    ext = archive_write_disk_new();
+    archive_write_disk_set_options(ext, flags);
+    archive_write_disk_set_standard_lookup(ext);
+    read = archive_read_open_filename(ar, filename, 1024);
+
+    if (read != ARCHIVE_OK) {
+        g_set_error(error, PU_ERROR, PU_ERROR_ARCHIVE_READ,
+                    "Failed reading archive '%s'", filename);
+        return FALSE;
+    }
+
+    old_dest = g_get_current_dir();
+    g_chdir(dest);
```

Comment:
Check the return value in case `dest` does not exist.

Reply:
Done.